File: client/go/src/example/finger/finger.go
package finger

import (
	"fmt"
	"context"
	"biostar/service/finger"
	"google.golang.org/grpc"
)

type FingerSvc struct {
	client finger.FingerClient
}

func NewFingerSvc(conn *grpc.ClientConn) *FingerSvc {
	return &FingerSvc{
		client: finger.NewFingerClient(conn),
	}
}

func (s *FingerSvc) Scan(deviceID uint32, templateFormat finger.TemplateFormat, qualityThreshold uint32) ([]byte, uint32, error) {
	req := &finger.ScanRequest{
		DeviceID: deviceID,
		TemplateFormat: templateFormat,
		QualityThreshold: qualityThreshold,
	}

	resp, err := s.client.Scan(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot scan a fingerprint: %v\n", err)

		return nil, 0, err
	}

	return resp.GetTemplateData(), resp.GetQualityScore(), nil
}


func (s *FingerSvc) GetImage(deviceID uint32) ([]byte, error) {
	req := &finger.GetImageRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.GetImage(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the last fingerprint image: %v\n", err)

		return nil, err
	}

	return resp.GetBMPImage(), nil
}


func (s *FingerSvc) GetConfig(deviceID uint32) (*finger.FingerConfig, error) {
	req := &finger.GetConfigRequest{
		DeviceID: deviceID,
	}

	resp, err := s.client.GetConfig(context.Background(), req)

	if err != nil {
		fmt.Printf("Cannot get the fingerprint config: %v\n", err)

		return nil, err
	}

	return resp.GetConfig(), nil
}

File: client/python/example/finger/finger.py
import grpc

import finger_pb2_grpc
import finger_pb2


class FingerSvc:
  stub = None

  def __init__(self, channel): 
    try:
      self.stub = finger_pb2_grpc.FingerStub(channel)
    except grpc.RpcError as e:
      print(f'Cannot get the finger stub: {e}')
      raise

  def scan(self, deviceID, templateFormat, qualityThreshold):
    try:
      response = self.stub.Scan(finger_pb2.ScanRequest(deviceID=deviceID, templateFormat=templateFormat, qualityThreshold=qualityThreshold))
      return response.templateData
    except grpc.RpcError as e:
      print(f'Cannot scan a finger: {e}')
      raise

  def getImage(self, deviceID):
    try:
      response = self.stub.GetImage(finger_pb2.GetImageRequest(deviceID=deviceID))
      return response.BMPImage
    except grpc.RpcError as e:
      print(f'Cannot get the fingerprint image: {e}')
      raise

  def getConfig(self, deviceID):
    try:
      response = self.stub.GetConfig(finger_pb2.GetImageRequest(deviceID=deviceID))
      return response.config
    except grpc.RpcError as e:
      print(f'Cannot get the fingerprint config: {e}')
      raise
